src/docfx/build/xref/XrefMapLoader.cs

```diff
@@ -0,0 +1,84 @@
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+using System.Text.Json;
+
+namespace Microsoft.Docs.Build
+{
+    internal class XrefMapLoader
+    {
+        public static ListBuilder<(string, Lazy<IXrefSpec>)> Load(string filePath)
+        {
+            var result = new ListBuilder<(string, Lazy<IXrefSpec>)>();
+            var content = File.ReadAllBytes(filePath);
+
+            // TODO: cache this position mapping if xref map file not updated, reuse it
+            var xrefSpecPositions = GetXrefSpecPositions(content);
+
+            ParallelUtility.ForEach(xrefSpecPositions, item =>
+            {
+                using (var stream = File.Open(filePath, FileMode.Open, FileAccess.Read, FileShare.Read))
+                {
+                    var json = ReadJsonFragment(stream, item.start, item.end);
+                    result.Add((item.uid, new Lazy<IXrefSpec>(() => JsonUtility.Deserialize<ExternalXrefSpec>(json, filePath))));
+                }
+            });
+            return result;
+        }
+
+        private static List<(string uid, long start, long end)> GetXrefSpecPositions(ReadOnlySpan<byte> content)
+        {
+            var result = new List<(string uid, long start, long end)>();
+            var reader = new Utf8JsonReader(content, isFinalBlock: true, default);
+            string uid = null;
+            int startIndex = 0;
+            int endIndex = 0;
+            while (reader.Read())
+            {
+                switch (reader.TokenType)
+                {
+                    case JsonTokenType.PropertyName:
+                        if (reader.TextEquals(Encoding.UTF8.GetBytes("uid")) && reader.Read() && reader.TokenType == JsonTokenType.String)
+                        {
+                            uid = Encoding.UTF8.GetString(reader.ValueSpan);
+                        }
+                        break;
+                    case JsonTokenType.StartObject:
+                        startIndex = (int)reader.TokenStartIndex;
+                        break;
+                    case JsonTokenType.EndObject:
+                        if (uid != null)
+                        {
+                            endIndex = (int)reader.TokenStartIndex;
+                            result.Add((uid, startIndex, endIndex));
+                        }
+                        break;
+                }
+            }
+            return result;
+        }
+
+        private static string ReadJsonFragment(Stream stream, long start, long end)
+        {
+            var bytesRead = 0;
+            var bytesToRead = end - start + 1;
+            var bytes = new byte[bytesToRead];
+            stream.Position = start;
+
+            do
+            {
+                var read = stream.Read(bytes, bytesRead, (int)bytesToRead);
+                bytesRead += read;
+                bytesToRead -= read;
+            } while (bytesToRead > 0);
+            stream.Close();
+
```

Comment:
the stream will be closed by the using statement  in line 29, so no need to close here